NF: Uncouple the exporting dialog/logic from the activities.

TarekkMA opened the pull request:
## Purpose / Description
Uncouple the exporting dialog/logic from the activities. 

With this PR now any activity which wish to implement exporting will have to add only these 3 lines
```java
ActivityExportingDelegate mExportingDelegate;
......
mExportingDelegate = new ActivityExportingDelegate(this /*this is of type AnkiActivity*/);
......
mExportingDelegate.showExportDialog(args....);
```


## Fixes
Base for #8159

## Approach
I have used a combination of `FragmentFactory` and the activity communication API, to fully uncouple the exporting dialogs from their hosting activities.

## How Has This Been Tested?
Manual testing

## Learning (optional, can help others)
https://medium.com/capital-one-tech/android-fragmentfactory-75823af015fd
https://proandroiddev.com/the-defective-androidx-fragmentfactory-599b63879f35

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

TarekkMA commented:
@Arthur-Milchior, @david-allison-1  here is the new design.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/export/ActivityExportingDelegate.java:
```diff
@@ -170,12 +191,15 @@ private Collection getCol() {
 
 
-    private static class ExportListener extends TaskListenerWithContext<AnkiActivity, Void, Pair<Boolean, String>> {
+    private static class ExportTaskListener extends TaskListenerWithContext<AnkiActivity, Void, Pair<Boolean, String>> {
```
Why do you add "Task" in the class name ? It's not a task.

TarekkMA replied:
I meant it to be XYZTaskListener so when I move it to its own file. The name would point that this is a task listener

Arthur-Milchior replied:
This is not consistent with current code. All the remaining code uses `XYZListener` to listen to result of `XYZ`. Maybe it's not ideal or ambiguous, but at least I'd prefer consistency with existing code until we decide to rename all Listeners to TaskListener

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/export/ActivityExportingDelegate.java:
```diff
@@ -35,24 +39,26 @@
  * <p>
  * The calling activity must implement {@link ExportCompleteDialogListener} and can then forward any call the the exporting delegate
  */
-public class ActivityExportingDelegate <A extends AnkiActivity> implements ExportDialogListener, ExportCompleteDialogListener {
```
Why do you add it and then remove it in the same PR?

TarekkMA replied:
Was building of the previous commits.

1. Make dialogs interact with interfaces not activities directly
2. Move logic to the delegate while keeping the interface implementation still in the activity
3. Uncouple the activity entirely from communicating with dialogs and move everything to the delegate.


Should I make the transition directly?

Arthur-Milchior replied:
No, I guess it's fine. Tat was surprising. Thanks for the explanation

TarekkMA replied:
Will rework the commits and try to make them more clear

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/export/ActivityExportingDelegate.java:
```diff
@@ -150,11 +156,19 @@ public void saveExportFile(String path) {
 
 
-    public boolean exportToProvider(Intent intent, boolean deleteAfterExport) {
+    private void saveFileCallback(ActivityResult result) {
```
```java
boolean data = exportToProvider(result.getData(), true);
String message = (data)?R.string.export_save_apkg_successful : R.string.export_save_apkg_unsuccessful;
UIUtils.showSimpleSnackbar(mActivity, mActivity.getString(message), data);
```

Arthur-Milchior reviewed (commented):
Can you split "Uncouple the Exporting dialogs from their hosting activities." more ? It seems to contains a lot of changes at once, and I'm having a hard time reviewing it. I must admit I don't know factory really so it's not clear to me what's done here, and I hope a smaller commit could help

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/ActivityExportingDelegate.java:
```diff
@@ -0,0 +1,207 @@
+ * <p>
+ * The calling activity must implement {@link ExportCompleteDialogListener} and can then forward any call the the exporting delegate
+ */
+public class ActivityExportingDelegate <A extends AnkiActivity & ExportCompleteDialogListener> {
```
Just to enforce the activity to implement this interface. Insted of checking on runtime

Arthur-Milchior replied:
But why does it needs to implement it ? In the commit where you introduce this line, you don't use any function from this interface.

TarekkMA replied:
I don't use it directly, but the `ExportCompleteDialog` does use it (by casting the hosting activity to this interface). So the activity must implement this interface.  

https://github.com/ankidroid/Anki-Android/blob/1e5b8f5b254b748d7bcf35a448b377edf66e22a7/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ExportCompleteDialog.java#L45-L52

This is the kind of coupling I removed in the following commits.

TarekkMA commented:
Okay will try to split it up.

TarekkMA commented:
@Arthur-Milchior 
I have cleaned the commits and split some commits into smaller units.

I have added some explanation on why I use `FragmentFactory` in the commit message.

TarekkMA commented:
If the new design is okay I will push a couple more commits with some tests

TarekkMA commented:
Converted to draft for now since I'm tweeking the API of `ExtendedFragmentFactory`

TarekkMA commented:
@Arthur-Milchior when you get the time, can you review this PR?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/export/ActivityExportingDelegate.java:
```diff
@@ -0,0 +1,206 @@
+     * @param activity the calling activity (must implement {@link ExportCompleteDialogListener})
+     * @param collectionSupplier a predicate that supplies a collection instance
+     */
+    public ActivityExportingDelegate(AnkiActivity activity, Supplier<Collection> collectionSupplier) {
```
We do have a `CollectionGetter` interface defined by Arthur - this may or may not be useful here -implementer's choice

TarekkMA replied:
This is a good idea, I will try to change it.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/export/ActivityExportingDelegate.java:
```diff
@@ -0,0 +1,206 @@
+        mActivity = activity;
+        this.mCollectionSupplier = collectionSupplier;
+        final FragmentManager fragmentManager = activity.getSupportFragmentManager();
+        mDialogsFactory = new ExportDialogsFactory(this, this).attachToActivity(activity);
```
It's atypical to have constructors do work, are there alternate implementations?

I'm open to designs (if you feel it's OK here, that's fine too). 

I feel a static factory method would go well here, but could also see it on the activity/as a default method somewhere.

TarekkMA replied:
Can we have an `attach` method that will do the work in the constructor? in that way we separate the work from the construction.

david-allison-1 replied:
Could we do a static method: `ExportDialogsFactory.createInstance(this, this)` for example

TarekkMA replied:
seems good, should I do this in this PR?

TarekkMA replied:
Do you mean `ActivityExportingDelegate.createInstance(this, this)`?

If you are talking about `ExportDialogsFactory`, then the constructor doesn't do any work the `attachToActivity` is doing the work.

david-allison-1 replied:
You're absolutely correct (sorry! too many things going on). 

Would moving the constructor to a static method work for this use case?

TarekkMA replied:
Maybe we can treat it like the factory, splitting it into 2 steps. Constructor and attach method.

david-allison-1 replied:
Sure!

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/export/ExportDialogsFactory.java:
```diff
@@ -0,0 +1,69 @@
+        Class<? extends Fragment> cls = loadFragmentClass(classLoader, className);
+
+        if (cls == ExportDialog.class) {
+            return newExportDialog();
```
Discarding the variable here (instead of using a setter) seems unusual: I'm not too familiar with this pattern. Is this normal?

If this is the only actual instantiation of an `ExportDialog`, then it looks good, it'll take a few days to get used to.

TarekkMA replied:
Can you provide more details on what you mean by discarding the variable?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/export/ActivityExportingDelegate.java:
```diff
@@ -0,0 +1,206 @@
+package com.ichi2.anki.export;
```
copyright

david-allison-1 reviewed (commented):
Shouldn't have let this sit, let's get it moving

david-allison-1 commented:
I'll want to see this in, we need more refactoring of the DeckPicker, and this was good work in that direction.

TarekkMA commented:
I will try to get this going.

david-allison-1 reviewed (approved):
Looks great, thanks!

TarekkMA commented:
Thank you I can address your comments in other PRs (Constructor and CollectionGetter), and sorry for blocking this PR for so long.

david-allison-1 commented:
No apologies necessary, life takes priority.

Glad to see you're back :)

Arthur-Milchior reviewed (commented):
I'm having trouble getting my head around
Uncouple the Exporting dialogs from their hosting activities.
and
Move onActivityResult to the new API.

The commits moving code are allright by themselves. Maybe it may be worth putting them in separate PR so that we don't have to re-review them when reviewing the more complex commit.

Since David is already giving constructive feedback on the part of the complex part of the work, I'll wait until he ask for a second review.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DialogHandler.java:
```diff
@@ -103,11 +102,7 @@ public void handleMessage(Message msg) {
             int id = msgData.getInt("dialogType");
             String message = msgData.getString("dialogMessage");
             ((DeckPicker) mActivity.get()).showSyncErrorDialog(id, message);
-        } else if (msg.what == MSG_SHOW_EXPORT_COMPLETE_DIALOG) {
```
okay

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ExportCompleteDialog.java:
```diff
@@ -79,15 +79,4 @@ public String getNotificationMessage() {
-
-
-    @Override
-    public Message getDialogHandlerMessage() {
```
`AsyncDialogFragment#getDialogHandlerMessage()` is still used in code in `AnkiActivity`.

and the method is overrided in `DatabaseErrorDialog`, `MediaCheckDialog`, `SyncErrorDialog`

https://github.com/ankidroid/Anki-Android/blob/242e253ed0378ebd2cc5c66cc126bec4ad9766d2/AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java#L506

Arthur-Milchior replied:
Strange. I missed it. Thanks

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:
```diff
@@ -94,10 +94,8 @@
 import com.ichi2.anki.dialogs.DeckPickerBackupNoSpaceLeftDialog;
```
Yeah, those are typos 😅. Fixed.

mikehardy reviewed (approved):
I don't see any comments / concerns that weren't addressed and the code all appears fine - here we go :-) - thanks @TarekkMA !

TarekkMA commented:
Now #8159 is clear, I will try to get it working and undraft it.

mikehardy commented:
I know exactly what you mean - I think I've mentioned it but I made the (in hindsight, stupid, but useful?) decision to start with CardTemplateEditor as my first PR. Something like a year later I finally got it in after putting in a ridiculous (but all needed!) amount of infrastructure around it. Anyway, forward progress :-) :steam_locomotive:

mikehardy commented:
Hi there!

Apologies for the delay, and if you're already submitted something to OpenCollective for PRs merged in August 2021, you may feel free to ignore this and it will be processed shortly

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for August 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

(I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged for in the month of August)

Thanks!
